proof_ics23.go

```diff
@@ -0,0 +1,166 @@
+	if rightkey != nil {
+		nonexist.Right, err = createExistenceProof(tree, rightkey)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	proof := &ics23.CommitmentProof{
+		Proof: &ics23.CommitmentProof_Nonexist{
+			Nonexist: nonexist,
+		},
+	}
+	return proof, nil
+}
+
+func createExistenceProof(tree *ImmutableTree, key []byte) (*ics23.ExistenceProof, error) {
+	value, proof, err := tree.GetWithProof(key)
+	if err != nil {
+		return nil, err
+	}
+	if value == nil {
+		return nil, fmt.Errorf("cannot create ExistanceProof when Key not in State")
+	}
+	return convertExistenceProof(proof, key, value)
+}
+
+// convertExistenceProof will convert the given proof into a valid
+// existence proof, if that's what it is.
+//
+// This is the simplest case of the range proof and we will focus on
+// demoing compatibility here
+func convertExistenceProof(p *RangeProof, key, value []byte) (*ics23.ExistenceProof, error) {
+	if len(p.Leaves) != 1 {
+		return nil, fmt.Errorf("existence proof requires RangeProof to have exactly one leaf")
+	}
+	return &ics23.ExistenceProof{
+		Key:   key,
+		Value: value,
+		Leaf:  convertLeafOp(p.Leaves[0].Version),
+		Path:  convertInnerOps(p.LeftPath),
+	}, nil
+}
+
+func convertLeafOp(version int64) *ics23.LeafOp {
+	// this is adapted from iavl/proof.go:proofLeafNode.Hash()
+	prefix := aminoVarInt(0)
+	prefix = append(prefix, aminoVarInt(1)...)
+	prefix = append(prefix, aminoVarInt(version)...)
+
+	return &ics23.LeafOp{
+		Hash:         ics23.HashOp_SHA256,
+		PrehashValue: ics23.HashOp_SHA256,
+		Length:       ics23.LengthOp_VAR_PROTO,
+		Prefix:       prefix,
+	}
+}
+
+// we cannot get the proofInnerNode type, so we need to do the whole path in one function
+func convertInnerOps(path PathToLeaf) []*ics23.InnerOp {
+	steps := make([]*ics23.InnerOp, 0, len(path))
+
+	// lengthByte is the length prefix prepended to each of the sha256 sub-hashes
+	var lengthByte byte = 0x20
+
+	// we need to go in reverse order, iavl starts from root to leaf,
+	// we want to go up from the leaf to the root
+	for i := len(path) - 1; i >= 0; i-- {
+		// this is adapted from iavl/proof.go:proofInnerNode.Hash()
+		prefix := aminoVarInt(int64(path[i].Height))
+		prefix = append(prefix, aminoVarInt(path[i].Size)...)
+		prefix = append(prefix, aminoVarInt(path[i].Version)...)
+
+		var suffix []byte
+		if len(path[i].Left) > 0 {
+			// length prefixed left side
+			prefix = append(prefix, lengthByte)
+			prefix = append(prefix, path[i].Left...)
+			// prepend the length prefix for child
+			prefix = append(prefix, lengthByte)
+		} else {
+			// prepend the length prefix for child
+			prefix = append(prefix, lengthByte)
+			// length-prefixed right side
+			suffix = []byte{lengthByte}
+			suffix = append(suffix, path[i].Right...)
+		}
+
+		op := &ics23.InnerOp{
+			Hash:   ics23.HashOp_SHA256,
+			Prefix: prefix,
+			Suffix: suffix,
+		}
+		steps = append(steps, op)
+	}
+	return steps
+}
+
+func aminoVarInt(orig int64) []byte {
```

Comment:
Is this really necessary? As far as I know, Amino uses standard Go variants, and we have utility functions for generating these in `encoding.go` - can you check if we can use any of those instead?